### PR TITLE
Adding attention block

### LINKED ISSUE #723: Add recurrent attention to RNN Block
`feature_request`.

### Feature Description

The current implementation adds plain attention to the inputs or the final outputs. Need recurrent attention on each layer of the RNN model

### Reason

Recurrent attention captures chronologic patterns in the representations learned by each RNN layer.

### PR DESCRIPTION
**Pull request by achaar**
Feature Description

The current implementation adds plain attention to the inputs or the final outputs. Need recurrent attention on each layer of the RNN model

resolves #723

### Details of the Pull Request

Previous code changes have been added to basic.py for attention in RNNBlock.

### CODE REVIEW

**Comment by achaar** [comment]
The RNN attention block has been added. Although, I am not sure how we can prevent it from being used in Bidirectional RNNs.

File: autokeras/blocks/basic.py [hunk 1]
```diff
@@ -154,6 +154,54 @@ def build(self, hp, inputs=None):
...
+    # Arguments
+        return_sequences: Boolean. Whether to return the last output in the
+            output sequence, or the full sequence. Defaults to False.
+    """
+
+    def __init__(self,
+                 return_sequences: bool = False,
+                 **kwargs):
+        super().__init__(**kwargs)
+        self.return_sequences = return_sequences
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:169]: I don't think we need this. If it is False, there is no need to use this attention block.

File: autokeras/blocks/basic.py [hunk 2]
```diff
@@ -154,6 +154,48 @@ def build(self, hp, inputs=None):
...
+
+    inputs: List of the following tensors:
+      * query: Query `Tensor` of shape `[batch_size, Tq, dim]`.
+      * value: Value `Tensor` of shape `[batch_size, Tv, dim]`.
+      * key: Optional key `Tensor` of shape `[batch_size, Tv, dim]`. If not
+        given, will use `value` for both `key` and `value`, which is the
+        most common case.
+
+    Output shape:
+    Attention outputs of shape `[batch_size, Tq, dim]`.
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:168]: I just realized it would be better to put this into the build function's docstring, as # Arguments and # Returns

**Comment by achaar** [comment]
I can present the attention block in tomorrow's meeting. But for it to work with the RNN block, the RNN block also needs to be modified. Currently, it does not return the hidden state of the RNN, which is required for the attention block.

**Comment by haifeng-jin** [comment]
@achaar 
I think there are two kinds of attentions.
One is Attention layer.
The other on is AdditiveAttention layer.
The latter one is more commonly used.
The original implementation in another pull request is about the latter one.

So we may discuss the two attention mechanisms and decide which one to use. Or may be we can use both.
You can find the paper links of these two in https://buomsoo-kim.github.io/attention/2020/01/01/Attention-mechanism-1.md/#:~:text=Attention%20Mechanism%20in%20Neural%20Networks%20%2D%201.&text=Attention%20is%20arguably%20one%20of,a%20large%20amount%20of%20information

Thanks.

**Comment by haifeng-jin** [comment]
@achaar I just consulted Keras team for how to do the attentions. It is not easy to implement. So let's just do the following.

1. Address the comments for moving the docstrings.
2. Use the hp to select between the two kinds of attention layers (Attention, AdditiveAttention).

Then this PR is good to be merged.
This Attention block will be used by users who are building their own search spaces with the blocks.

We will explore other ways to improve the performance of the text tasks.

Thanks.

File: autokeras/blocks/basic.py [hunk 3]
```diff
@@ -154,6 +154,68 @@ def build(self, hp, inputs=None):
...
+        super().__init__(**kwargs)
+
+    def get_config(self):
+        config = super().get_config()
+        config.update({
+            'attention_type': self.attention_type})
+        return config
+
+    def build(self, hp, inputs=None):
+        '''
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:179]: Should use """

File: autokeras/blocks/basic.py [hunk 4]
```diff
@@ -154,6 +154,68 @@ def build(self, hp, inputs=None):
...
+
+        inputs: List of the following tensors:
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:181]: Put these under # Arguments

File: autokeras/blocks/basic.py [hunk 5]
```diff
@@ -154,6 +154,68 @@ def build(self, hp, inputs=None):
         return output_node
 
 
+class AttentionBlock(block_module.Block):
+    """The Attention Block. Luong's Multiplicative Style Attention.
+
+    # Arguments
+        attention_type: String. 'multiplicative' or 'additive'.
+        If left unspecified, it will be tuned automatically.
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:162]: indent

File: autokeras/blocks/basic.py [hunk 6]
```diff
@@ -154,6 +154,68 @@ def build(self, hp, inputs=None):
...
+        '''
+
+        inputs: List of the following tensors:
+          * query: Query `Tensor` of shape `[batch_size, Tq, dim]`.
+          * value: Value `Tensor` of shape `[batch_size, Tv, dim]`.
+          * key: Optional key `Tensor` of shape `[batch_size, Tv, dim]`. If not
+            given, will use `value` for both `key` and `value`, which is the
+            most common case.
+
+        Output shape:
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:188]: # Returns

File: autokeras/blocks/basic.py [hunk 7]
```diff
@@ -154,6 +154,69 @@ def build(self, hp, inputs=None):
         return output_node
 
 
+class AttentionBlock(block_module.Block):
+    """The Attention Block. Luong's Multiplicative Style Attention.
+
+    # Arguments
+        attention_type: String. 'multiplicative' or 'additive'.
+            If left unspecified, it will be tuned automatically.
+
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:163]: Remove this empty line.

File: autokeras/blocks/basic.py [hunk 8]
```diff
@@ -154,6 +154,69 @@ def build(self, hp, inputs=None):
...
+            inputs: List of the following tensors:
+              * query: Query `Tensor` of shape `[batch_size, Tq, dim]`.
+              * value: Value `Tensor` of shape `[batch_size, Tv, dim]`.
+              * key: Optional key `Tensor` of shape `[batch_size, Tv, dim]`. If not
+                given, will use `value` for both `key` and `value`, which is the
+                most common case.
+
+        # Returns:
+            Attention outputs of shape `[batch_size, Tq, dim]`.
+
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:191]: Remove this empty line.

File: autokeras/blocks/basic.py [hunk 9]
```diff
@@ -154,6 +154,69 @@ def build(self, hp, inputs=None):
...
+            'additive': layers.AdditiveAttention
+        }
+        if len(inputs) == 2 or len(inputs) == 3:
+            # Either [query, value] or [query, value, key] provided
+            for in_vec in inputs:
+                if len(in_vec.shape) != 3:
+                    raise ValueError(
+                        'Expect 3 dimensions for each of query, value, (or key) '
+                        'but got {shape}'.format(shape=in_vec.shape)
+                    )
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:209]: I expect we have this check in the Keras Attention layers. You may have a try to see the exception the Attention and AdditiveAttention raises. If they have, we can remove this check.

File: autokeras/blocks/basic.py [hunk 10]
```diff
@@ -154,6 +154,69 @@ def build(self, hp, inputs=None):
...
+            output_node = attention_layers[attention_type](inputs)
+        else:
+            raise ValueError(
+                'Expect the input to be [query, value] or [query, value, key]'
+                'for the Attention Layer, '
+                'but got {shape}'.format(shape=inputs.shape))
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:215]: The same for this error. I expect the keras layer checked this.

File: autokeras/blocks/basic.py [hunk 11]
```diff
@@ -154,6 +154,69 @@ def build(self, hp, inputs=None):
         return output_node
 
 
+class AttentionBlock(block_module.Block):
+    """The Attention Block. Luong's Multiplicative Style Attention.
```
Comment (haifeng-jin) [inline autokeras/blocks/basic.py:158]: This description needs to add the AddtiveAttention layer.
Just change this line to "Using Attention or AdditiveAttention layer in Keras."

File: autokeras/blocks/basic.py (hunk 9 above)
Comment (haifeng-jin) [reply autokeras/blocks/basic.py:209]: If they don't have the exception. We need to check each tensor respectively and give the expected dimensions in details like [batch_size, Tq, dim] in the error message.

**Pull request closed without merging**